src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,443 @@
+        /// computing the model to know what to filter it down to.</remarks>
+        public string SearchText
+        {
+            get
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                return _searchText_doNotAccessDirectly;
+            }
+
+            set
+            {
+                // Called from WPF.  When this changes, kick off the work to actually filter down our models.
+
+                _threadingContext.ThrowIfNotOnUIThread();
+                _searchText_doNotAccessDirectly = value;
+
+                _workQueue.AddWork(default(VoidResult), cancelExistingWork: true);
+            }
+        }
+
+        /// <remarks>This property is bound to the UI.  It is only read by the UI, but can be read/written by us.</remarks>
+        public ImmutableArray<DocumentSymbolDataViewModel> DocumentSymbolViewModelItems
+        {
+            get
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                return _documentSymbolViewModelItems_doNotAccessDirectly;
+            }
+
+            // Setting this only happens from within this type once we've computed new items or filtered down the existing set.
+            private set
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                if (_documentSymbolViewModelItems_doNotAccessDirectly == value)
+                    return;
+
+                _documentSymbolViewModelItems_doNotAccessDirectly = value;
+                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(DocumentSymbolViewModelItems)));
+            }
+        }
+
+        public void ExpandOrCollapseAll(bool shouldExpand)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+            ExpandOrCollapse(this.DocumentSymbolViewModelItems, shouldExpand);
+
+            static void ExpandOrCollapse(ImmutableArray<DocumentSymbolDataViewModel> models, bool shouldExpand)
+            {
+                foreach (var model in models)
+                {
+                    model.IsExpanded = shouldExpand;
+                    ExpandOrCollapse(model.Children, shouldExpand);
+                }
+            }
+        }
+
+        private async ValueTask ComputeViewStateAsync(CancellationToken cancellationToken)
+        {
+            // We do not want this work running on a background thread
+            await TaskScheduler.Default;
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Do the expensive LSP work of actually computing the items to show.
+            var (documentSymbolData, newTextSnapshot) = await ComputeDocumentSymbolDataAsync(cancellationToken).ConfigureAwait(false);
+
+            // Now, go back to the UI and grab the prior view state we set, and the current UI values we want to update the data with.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
This is _deeply_ unfortunate. I would like us to do some measurements under perfview make sure that we do not see [AzDo#1678321](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1678321) recurr. What this means is that when Visual Studio first opens a project, if Document Outline is open, we are fighting to get on the UI thread. What is the reason for switching to the UI thread? Do we expect multiple threads to be running `ComputeViewStateAsync` concurrently? Otherwise I do not understand this requirement.